### PR TITLE
Feature/achievements widget

### PR DESCRIPTION
**Pull request by lfortemps**
### What does it do?

Initial contribution to #5713 . It brings gamification through an 'achievements' widget which can help new users discover the different MISP features. An example is shown below: 

achievements_widget

I'm planning to add more badges but would like to request an initial feedback, what do you think?

#### Questions

- [ ] Does it require a DB change?
- [ ] Are you using it in production?
- [ ] Does it require a change in the API (PyMISP for example)?

### CODE REVIEW

**Comment by cvandeplas** [comment]
cool  ! 

- A first visual impression is the text size which is imbalanced with the icons. (so increase text size or shrink images, or both?)
- For the icons you could use the MISP Object and MISP Event icons : MISPEvent96 and MISPObject96
- Taxonomies:  taxonomies or taxonomies 
- Galaxies: galaxy or  galaxy
- Relationships: relation or  relation

File: app/Lib/Dashboard/AchievementsWidget.php [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+{
+  /*
+  * Note: for this widget to display as expected, you need all icons to be accessible in your webroot. (img/custom)
+  * Icons used:
+  * - relationship.png --> https://raw.githubusercontent.com/MISP/intelligence-icons/master/simple_png/64/relationship.png
+  * - report.png --> https://raw.githubusercontent.com/MISP/intelligence-icons/master/square_png/64/report.png
+  * - tlp_green.png --> https://raw.githubusercontent.com/MISP/intelligence-icons/master/square_png/64/tlp_green.png
+  */
+
+	public $render = 'Achievements';
```
Comment (SteveClement) [inline app/Lib/Dashboard/AchievementsWidget.php]: I replaced the tabs with spaces.

**Comment by lfortemps** [comment]
Slightly modified visually to take into account @cvandeplas comment. Also added 3 new badges.
achievements_widget2

**Comment by cvandeplas** [comment]
To speed up the query, would using a `LIMIT 0,1` type of filter be a good idea? This way the DB can stop once it has found one matching item. 
If that speeds up things, we could also test removing the time-filter as logically it would slow down things again.

**Comment by lfortemps** [comment]
As I'm using a "count" instead of a regular select, queries are really fast. At least on my small DB. I would have to test on a large DB. Why would you want to remove the time limit? If people go inactive for a long time, it's good to remotivate them :)

**Comment by cvandeplas** [comment]
Excellent point on the time limit,
Could you do a benchmark on the `count` vs `limit 1` on prod data? Eager to see the effect on performance.

**Comment by mokaddem** [comment]
I love the idea!
How do you plan to make it visible to the users?
Make their favorite page the 'achievement dashboard' page?

**Comment by mokaddem** [comment]
He are some more ideas and remarks we received for the [MISP dashboard](https://github.com/MISP/misp-dashboard) project that could also be interesting here:

- Specific badges and achievements for interacting with specific threat actors
- Rewards for discovering APT 41 C2, or new important TTPs
- Rewards for industry best champion (who produces the most/highest quality threat intel on actors targeting our industry)
    - Just be careful with the **most** as you want to avoid people submitting bad quality/junk content
- Better cross industry trends, filtering by industry
- Rewards or points for intra-industry collaboration, rather than analyst output

Not sure how to implement these point in a clean way, so food for thought

Credit @VVX7 ;)

**Comment by lfortemps** [comment]
@cvandeplas 
So as we could expect the larger the number of events, the better the "limit 1" will perform. For an organization having 20k events, this could go up to 0.4s via count and 0.01s for limit 1.
However for organizations having few events, the count completes faster (1ms) vs 10ms for the limit 1.  
So it's best if I switch the sql statements to select with limit instead of count.

@mokaddem Yes the plan is to put it in a dashboard that would be the default page after log in, next to potentially other widgets.
Thanks for the ideas, it might come in a subsequent PR but indeed it needs to be thought beforehand to make sure it doesn't introduce a huge computation cost and also define the appropriate weight/score on each contribution.

**Comment by cvandeplas** [comment]
I' ve added @mokaddem 's feedback in the feature request: https://github.com/MISP/MISP/issues/5713#issuecomment-662441777

**Comment by cvandeplas** [comment]
thank you very much ! 
Is there something else needed (SQL) to add it as part of the default dashboard?

**Comment by lfortemps** [comment]
I have checked and there are a couple of problems with making this the default for all MISP install:

- It is indeed stored in SQL, so we would need to modify the INSTALL/MySQL.sql but there would be a broken reference because no user_id (template creator) can be populated at that time. Or we can make it part of a DB migration step? but again how do we decide on the user_id
- The pictures would need to be added in the repository by default
- Some people might not like to have it accessible for all orgs in their instance
- The layout can be great for our screen and instance data but it might not suit someone's else instance

So my vote would be to let MISP admin chose to create a template with the widgets available instead of forcing a default template. What do you guys think?

**Comment by cvandeplas** [comment]
thank you very much !

**Pull request merged**